jfix-stdlib-batching/src/main/java/ru/fix/stdlib/batching/BatchingManagerThread.java

```diff
@@ -112,102 +110,159 @@ public void run() {
-                            .map(Operation::getCreationTimestamp)
-                            .min(Long::compareTo);
-
-                    long waitTime;
-                    if (oldestCreationTimestatmp.isPresent()) {
-                        waitTime = batchingParameters.getBatchTimeout() - (System.currentTimeMillis() -
-                                oldestCreationTimestatmp.get());
-                        if (waitTime < 0) {
-                            waitTime = 1;
-                        }
-                    } else {
-                        waitTime = batchingParameters.getBatchTimeout();
-                    }
-                    log.trace("queue is empty, wait {}ms", waitTime);
-                    synchronized (waitForOperationLock) {
-                        try {
-                            if (waitTime <= 0) {
-                                waitTime = 1;
-                            }
-                            waitForOperationLock.wait(waitTime);
-                            log.trace("leave wait section");
-                        } catch (InterruptedException exc) {
-                            log.trace(THREAD_INTERRUPTED_MESSAGE, exc);
-                            Thread.currentThread().interrupt();
-                            return;
-                        }
-                    }
+                    if (!calculateTimeForSleepAndWaitForOperations()) return;
                 }
             }
 
             if (!isShutdown.get() && buffer != null) {
                 log.trace("Start batch processor");
-                batchProcessorPool.execute(new BatchProcessor<>(config,
+                executeBatchForBuffer(buffer, key);
+            }
+        }
+
+        metricsProvider.detachOperationQueueSizeIndicators();
+        log.trace("BatchProcessorManager thread stopped.");
+    }
+
+    private void executeBatchForBuffer(List<Operation<PayloadT>> buffer, KeyT key) {
+        batchProcessorPool.execute(
+                new BatchProcessor<>(
+                        config,
                         buffer,
                         batchProcessorsTracker,
                         batchTask,
                         key,
                         batchManagerId,
-                        profiler));
+                        profiler
+                )
+        );
+    }
+
+    private boolean waitForAvailableBatchProcessThread() {
+        log.trace("Wait available thread in BatchProcessor thread pool");
+        boolean isBatchProcessorThreadAvailable = false;
+        metricsProvider.startBatchProcessorThreadAwaitProfiling();
+        while (!isShutdown.get() && !isBatchProcessorThreadAvailable) {
+            try {
+                isBatchProcessorThreadAvailable = batchProcessorsTracker.tryAcquire(100,
+                        TimeUnit.MILLISECONDS);
+                if (isBatchProcessorThreadAvailable) {
+                    log.trace("isBatchProcessorThreadAvailable");
```

Comment:
Let's add batchManagerId to the log. in our project there are two instances.

Reply:
Done :)